File: dlgedit/DlgEditEvents.go
//go:build windows

package dlgedit

import (
	"fmt"
	"id3fit/ids"
	"slices"
	"strings"

	"github.com/rodrigocfd/windigo/co"
	"github.com/rodrigocfd/windigo/ui"
)

func (me *DlgEdit) events() {

	me.wnd.On().WmInitDialog(func(_ ui.WmInitDialog) bool {
		me.wnd.Hwnd().SetWindowText(fmt.Sprintf("Editing %d ID3v2 tag(s)", len(me.tags)))

		me.lstFrames.SetExtendedStyle(true, co.LVS_EX_FULLROWSELECT|co.LVS_EX_GRIDLINES)
		me.lstFrames.Cols.Add("Frame", ui.DpiX(56))
		me.lstFrames.Cols.Add("Value", ui.DpiX(100))

		me.fillComboGenres()
		me.fillTextboxes()
		me.fillFramesList()
		pixels, nBytes := me.wndPic.LoadPicture(me.tags)
		me.fillPicInfo(pixels, nBytes)

		return true
	})

	me.wnd.On().WmInitMenuPopup(func(p ui.WmInitMenuPopup) {
		firstId, _ := p.HMenu().GetMenuItemID(0)
		if firstId == ids.MNU_FRAMES_MOVEUP {
			oneTag := len(me.tags) == 1
			hasSel := me.lstFrames.Items.SelectedCount() > 0
			firstItem := me.lstFrames.Items.Get(0)
			lastItem := me.lstFrames.Items.Last()

			p.HMenu().EnableMenuItemByCmd(oneTag && hasSel && !firstItem.IsSelected(), ids.MNU_FRAMES_MOVEUP)
			p.HMenu().EnableMenuItemByCmd(oneTag && hasSel && !lastItem.IsSelected(), ids.MNU_FRAMES_MOVEDOWN)
			p.HMenu().EnableMenuItemByCmd(oneTag && hasSel, ids.MNU_FRAMES_DELETE)
		}
	})

	for _, input := range me.inputs {
		input.chk.On().BnClicked(func() { // checkboxes enable/disable inputs
			if input.chk.IsChecked() {
				input.txt.Hwnd().EnableWindow(true)
				input.txt.Focus()
			} else {
				input.txt.Hwnd().EnableWindow(false)
			}
		})
	}

	me.btnUncheckAll.On().BnClicked(func() {
		for _, input := range me.inputs {
			input.chk.SetCheckAndTrigger(false) // uncheck everyone
		}
	})

	me.btnCheckFilled.On().BnClicked(func() {
		for _, input := range me.inputs {
			txt, _ := input.txt.Hwnd().GetWindowText()
			if strings.TrimSpace(txt) != "" {
				input.chk.SetCheckAndTrigger(true) // check if has text
			}
		}
	})

	me.wnd.On().WmCommandAccelMenu(uint16(co.ID_OK), func() {
		me.writeTextsToTags()
		me.result = co.ID_OK
		me.wnd.Hwnd().SendMessage(co.WM_CLOSE, 0, 0)
	})

	me.wnd.On().WmCommandAccelMenu(uint16(co.ID_CANCEL), func() {
		me.result = co.ID_CANCEL
		me.wnd.Hwnd().SendMessage(co.WM_CLOSE, 0, 0)
	})

	me.wnd.On().WmCommandAccelMenu(ids.MNU_FRAMES_MOVEUP, func() {
		focusedItem, hasFocused := me.lstFrames.Items.Focused()

		selItems := me.lstFrames.Items.Selected()
		for _, sel := range selItems {
			idx := sel.Index()
			me.tags[0].Frames()[idx], me.tags[0].Frames()[idx-1] =
				me.tags[0].Frames()[idx-1], me.tags[0].Frames()[idx]
		}
		me.fillFramesList()
		for _, sel := range selItems {
			prev, _ := sel.Prev()
			prev.Select(true)
		}

		if hasFocused {
			me.lstFrames.Items.Get(focusedItem.Index() - 1).Focus()
		}
	})

	me.wnd.On().WmCommandAccelMenu(ids.MNU_FRAMES_MOVEDOWN, func() {
		focusedItem, hasFocused := me.lstFrames.Items.Focused()

		selItems := me.lstFrames.Items.Selected()
		for _, selItem := range slices.Backward(selItems) {
			idx := selItem.Index()
			me.tags[0].Frames()[idx], me.tags[0].Frames()[idx+1] =
				me.tags[0].Frames()[idx+1], me.tags[0].Frames()[idx]
		}
		me.fillFramesList()
		for _, selItem := range selItems {
			next, _ := selItem.Next()
			next.Select(true)
		}

		if hasFocused {
			me.lstFrames.Items.Get(focusedItem.Index() + 1).Focus()
		}
	})

	me.wnd.On().WmCommandAccelMenu(ids.MNU_FRAMES_DELETE, func() {
		selItems := me.lstFrames.Items.Selected()
		text := fmt.Sprintf("Do you want to remove %d frame(s)?", len(selItems))

		if ui.MsgOkCancel(me.wnd, "Remove frames", "", text, "&Remove") {
			for _, selItem := range slices.Backward(selItems) {
				me.tags[0].RemoveFrame(selItem.Index())
			}
			me.fillFramesList()
		}
	})

}

File: src/dlgs/dlg_edit_events.rs
use winsafe::{self as w, cmd, co, evp, evt, gui, prelude::*};

use super::{DlgEdit, ids};
use crate::msgbox;

impl DlgEdit {
	pub(super) fn events(&self) {
		evp!(self, wm_init_dialog, on_init_dialog);
		evp!(self, wm_init_menu_popup, on_init_menu_popup);
		cmd!(self, ids::MNU_FRAMES_MOVEUP, on_menu_frames_moveup);
		cmd!(self, ids::MNU_FRAMES_MOVEDOWN, on_menu_frames_movedown);
		cmd!(self, ids::MNU_FRAMES_DELETE, on_menu_frames_delete);
		evp!(self, self.lst_frames, lvn_key_down, on_lstframes_keydown);
		evt!(self, self.chk_pic, bn_clicked, on_chk_pick);
		evt!(self, self.btn_uncheck_all, bn_clicked, on_btn_unckeck_all);
		evt!(self, self.btn_check_filled, bn_clicked, on_btn_check_filled);
		cmd!(self, co::DLGID::OK, on_ok);
		cmd!(self, co::DLGID::CANCEL, on_esc);

		self.inputs.iter().for_each(|input| {
			let input2 = input.clone();
			input.chk.on().bn_clicked(move || {
				if input2.chk.is_checked() {
					input2.txt.hwnd().EnableWindow(true);
					input2.txt.focus()?; // when checked, enable the textbox and focus it
				} else {
					input2.txt.hwnd().EnableWindow(false);
				}
				Ok(())
			});
		});
	}

	fn on_init_dialog(&self, _: w::msg::WmInitDialog) -> w::AnyResult<bool> {
		// Add all the genres to the ComboBox.
		self.load_combo_genres()?;

		// For each checkbox + input, check if the frame has the same value across all tags.
		{
			let sel_tags = self.sel_tags.try_borrow()?;
			self.inputs
				.iter()
				.try_for_each(|input| input.set_text_if_equal_in_tags(&sel_tags))?;
		}

		// Setup the frames listview.
		let lv = &self.lst_frames;
		lv.cols().add("Frame", gui::dpi_x(56))?;
		lv.cols()
			.add("Value", gui::dpi_x(100))?
			.set_width_to_fill()?;
		lv.set_extended_style(true, co::LVS_EX::FULLROWSELECT | co::LVS_EX::GRIDLINES);

		self.render_frames_list()?;
		self.load_picture()?;
		self.chk_pic.hwnd().EnableWindow(false); // to be implemented later
		Ok(true)
	}

	fn on_init_menu_popup(&self, p: w::msg::WmInitMenuPopup) -> w::AnyResult<()> {
		let lv = &self.lst_frames;
		if p.hmenu == lv.context_menu().unwrap() {
			let has_sel = lv.items().selected_count() >= 1;
			let first_is_sel = lv.items().get(0).is_selected();
			let last_is_sel = lv.items().last().unwrap().is_selected();

			p.hmenu
				.EnableMenuItem(w::IdPos::Id(ids::MNU_FRAMES_MOVEUP), has_sel && !first_is_sel)?;
			p.hmenu
				.EnableMenuItem(w::IdPos::Id(ids::MNU_FRAMES_MOVEDOWN), has_sel && !last_is_sel)?;
			p.hmenu
				.EnableMenuItem(w::IdPos::Id(ids::MNU_FRAMES_DELETE), has_sel)?;
		}
		Ok(())
	}

	fn on_menu_frames_moveup(&self) -> w::AnyResult<()> {
		let items = self.lst_frames.items();
		let focus_idx = items.focused().map(|item| item.index());

		let new_sel_indexes = items // indexes of items that will appear as selected after the move
			.iter_selected()
			.map(|sel_item| {
				let idx = sel_item.index() as usize;
				self.sel_tags.try_borrow_mut()?[0] // assume we have only 1 MP3 loaded
					.frames_mut()
					.swap(idx, idx - 1); // swap frames in tag
				Ok(idx - 1)
			})
			.collect::<w::AnyResult<Vec<_>>>()?;

		self.render_frames_list()?;

		items.select_all(false)?;
		new_sel_indexes
			.iter()
			.try_for_each(|new_sel_idx| -> w::AnyResult<_> {
				items.get(*new_sel_idx as _).select(true)?; // re-select the moved items
				Ok(())
			})?;

		focus_idx.map(|idx| items.get(idx - 1).focus());
		Ok(())
	}

	fn on_menu_frames_movedown(&self) -> w::AnyResult<()> {
		let items = self.lst_frames.items();
		let focus_idx = items.focused().map(|item| item.index());

		let new_sel_indexes = items // indexes of items that will appear as selected after the move
			.iter_selected()
			.rev()
			.map(|sel_item| {
				let idx = sel_item.index() as usize;
				self.sel_tags.try_borrow_mut()?[0] // assume we have only 1 MP3 loaded
					.frames_mut()
					.swap(idx, idx + 1); // swap frames in tag
				Ok(idx + 1)
			})
			.collect::<w::AnyResult<Vec<_>>>()?;

		self.render_frames_list()?;

		items.select_all(false)?;
		new_sel_indexes
			.iter()
			.try_for_each(|new_sel_idx| -> w::AnyResult<_> {
				items.get(*new_sel_idx as _).select(true)?; // re-select the moved items
				Ok(())
			})?;

		focus_idx.map(|idx| items.get(idx + 1).focus());
		Ok(())
	}

	fn on_menu_frames_delete(&self) -> w::AnyResult<()> {
		if self.lst_frames.items().selected_count() == 0 {
			// Must check because listview's keydown will pop here even if
			// there are no selected items.
			return Ok(());
		}

		let msg = format!("Delete {} selected frame(s)?", self.lst_frames.items().selected_count());
		if msgbox::ask(&self.wnd, "Delete frame(s)", None, &msg, "&Delete")? {
			self.lst_frames.items().iter_selected().rev().try_for_each(
				|sel_item| -> w::AnyResult<_> {
					// Remove the frame by index directly from tag.
					self.sel_tags.try_borrow_mut()?[0] // assume we have only 1 MP3 loaded
						.frames_mut()
						.remove(sel_item.index() as _);
					Ok(())
				},
			)?;

			self.render_frames_list()?;
		}
		Ok(())
	}

	fn on_lstframes_keydown(&self, p: &w::NMLVKEYDOWN) -> w::AnyResult<()> {
		if p.wVKey == co::VK::DELETE {
			self.wnd
				.hwnd()
				.SendCommand(w::AccelMenuCtrl::Menu(ids::MNU_FRAMES_DELETE));
		}
		Ok(())
	}

	fn on_chk_pick(&self) -> w::AnyResult<()> {
		if self.chk_pic.is_checked() {
			self.wnd_pic.load_picture(&self.sel_tags.try_borrow()?)?; // ask the control to load the IPicture
		} else {
			self.wnd_pic.unload_picture()?;
		}
		Ok(())
	}

	fn on_btn_unckeck_all(&self) -> w::AnyResult<()> {
		self.inputs
			.iter()
			.try_for_each(|input| input.chk.set_check_and_trigger(false))?;
		self.chk_pic.set_check(false);
		Ok(())
	}

	fn on_btn_check_filled(&self) -> w::AnyResult<()> {
		self.inputs
			.iter()
			.try_for_each(|input| -> w::AnyResult<_> {
				let text = input.txt.hwnd().GetWindowText()?;
				if !text.is_empty() {
					input.chk.set_check(true);
					input.txt.hwnd().EnableWindow(true);
				}
				Ok(())
			})?;

		let has_pic = self.wnd_pic.pic.try_borrow()?.is_some();
		self.chk_pic.set_check(has_pic);
		Ok(())
	}

	fn on_ok(&self) -> w::AnyResult<()> {
		let mut sel_tags = self.sel_tags.try_borrow_mut()?;

		if sel_tags.len() == 1 {
			// If we're editing a single MP3 file, replace the frames with the
			// current ones in the listview.
			sel_tags[0].frames_mut().clear(); // delete all frames
			self.lst_frames
				.items()
				.iter()
				.try_for_each(|item| -> w::AnyResult<_> {
					let rc_frame = item.data();
					let cloned_frame = rc_frame.try_borrow()?.clone();
					sel_tags[0].frames_mut().push(cloned_frame); // add the frame from the listview
					Ok(())
				})?;
		}

		// Run through all textboxes and set/remove the text values on all tags.
		self.inputs
			.iter()
			.filter(|input| input.chk.is_checked())
			.try_for_each(|input| -> w::AnyResult<_> {
				let text = input.txt.hwnd().GetWindowText()?;
				sel_tags
					.iter_mut()
					.try_for_each(|tag| tag.set_editable_string(&input.name4, &text))?;
				Ok(())
			})?;

		self.user_clicked_ok.set(true);
		self.wnd.close();
		Ok(())
	}

	fn on_esc(&self) -> w::AnyResult<()> {
		self.user_clicked_ok.set(false);
		self.wnd.close();
		Ok(())
	}
}
